syntax = "proto3";

package second.protos;

// Messages for configuring the optimizing strategy for training object
// detection models.

// Top level optimizer message.
message Optimizer {
  oneof optimizer {
    RMSPropOptimizer rms_prop_optimizer = 1;
    MomentumOptimizer momentum_optimizer = 2;
    AdamOptimizer adam_optimizer = 3;
  }
  bool use_moving_average = 4;
  float moving_average_decay = 5;
  bool fixed_weight_decay = 6; // i.e. AdamW
}

// Configuration message for the RMSPropOptimizer
// See: https://www.tensorflow.org/api_docs/python/tf/train/RMSPropOptimizer
message RMSPropOptimizer {
  LearningRate learning_rate = 1;
  float momentum_optimizer_value = 2;
  float decay = 3;
  float epsilon = 4;
  float weight_decay = 5;
}

// Configuration message for the MomentumOptimizer
// See: https://www.tensorflow.org/api_docs/python/tf/train/MomentumOptimizer
message MomentumOptimizer {
  LearningRate learning_rate = 1;
  float momentum_optimizer_value = 2;
  float weight_decay = 3;
}

// Configuration message for the AdamOptimizer
// See: https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer
message AdamOptimizer {
  LearningRate learning_rate = 1;
  float weight_decay = 2;
  bool amsgrad = 3;
}

message LearningRate {
  oneof learning_rate {
    MultiPhase multi_phase = 1;
    OneCycle one_cycle = 2;
    ExponentialDecay exponential_decay = 3;
    ManualStepping manual_stepping = 4;
    ExponentialDecayWarmup exponential_decay_warmup=5;
  }
}

message LearningRatePhase {
  float start = 1;
  string lambda_func = 2;
  string momentum_lambda_func = 3;
}

message MultiPhase {
  repeated LearningRatePhase phases = 1;
}

message OneCycle {
  float lr_max = 1;
  repeated float moms = 2;
  float div_factor = 3;
  float pct_start = 4;
  repeated float lr_maxs = 5;

}

/*
ManualStepping example:
initial_learning_rate = 0.001
decay_length = 0.1
decay_factor = 0.8
staircase = True
detail:
progress 0%~10%, lr=0.001
progress 10%~20%, lr=0.001 * 0.8
progress 20%~30%, lr=0.001 * 0.8 * 0.8
......
*/


message ExponentialDecay {
  float initial_learning_rate = 1;
  float decay_length = 2; // must in range (0, 1)
  float decay_factor = 3;
  bool staircase = 4;
}

message ExponentialDecayWarmup {
  float initial_learning_rate = 1;
  float decay_length = 2; // must in range (0, 1)
  float decay_factor = 3;
  bool staircase = 4;
  float div_factor=6;
  float pct_start=7;
}

/*
ManualStepping example:
boundaries = [0.8, 0.9]
rates = [0.001, 0.002, 0.003]
detail:
progress 0%~80%, lr=0.001
progress 80%~90%, lr=0.002
progress 90%~100%, lr=0.003
*/

message ManualStepping {
  repeated float boundaries = 1; // must in range (0, 1)
  repeated float rates = 2;
}